### PR TITLE
chore: support for Helm 3

### LINKED ISSUE #127: Upgrade to Helm 3
Helm 2 is supported until Nov 2020

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
We are removing all code to support Helm 2, using Helm 3.2.1 as the target version

## Why is important?
Helm 2 is supported until Nov 2020

## Related issues
- Closes #127

### CODE REVIEW

**Comment by mdelapenya** [comment]
I'm going to keep this PR open until the Helm charts team confirms the move to Helm 3

cc/ @jmlrt

**Comment by mdelapenya** [comment]
It's interesting that the helm chart for APM Server fails in Helm 3 in this step:

```gherkin
A "Service" which will expose the pods as network services internal to the k8s cluster
```
with this error:
> yaml: line 7: mapping values are not allowed in this context

@jmlrt do you think this is a valid error?

**Comment by mdelapenya** [comment]
I found this: https://github.com/helm/helm/issues/6251, where https://github.com/helm/helm/issues/6251#issuecomment-622602014 seems relevant

wdyt about this line? https://github.com/elastic/helm-charts/blob/master/apm-server/templates/deployment.yaml#L122

**Comment by jmlrt** [comment]
Hi @mdelapenya 

> @jmlrt do you think this is a valid error?

I quickly tried to reproduce it locally but apm-server chart install is OK with Helm 3.2.1 and default values for me.

Can you give more details about this test:
- Are you using default values or overriding them?
- Do you have other log to help debug?

> I found this: helm/helm#6251, where helm/helm#6251 (comment) seems relevant

Regarding than some issues mentionned are happening to charts with `-` in the name and that apm-server is the only chart we have like that, it could be worse trying to reproduce your error while changing chart name in [Chart.yaml](https://github.com/elastic/helm-charts/blob/4eceebc18194b816f8439a03c4598d0f8359769b/apm-server/Chart.yaml#L7). Another test could be to replace `{{ .Chart.Name }}` by `{{ .Chart.Name | quote }}` in the different templates.

> I'm going to keep this PR open until the Helm charts team confirms the move to Helm 3

Current work for Helm 3 is tracked in https://github.com/elastic/helm-charts/pull/516. I had to reprioritize it and I hope being able to add Helm 3 support for 7.9.0 but that may be moved to 7.10 depending on the other priorities for release team (7.10 is planned for november 3rd so it would come a little late regarding Helm 2 end of support in november). 

Does that look fine for you?

**Comment by mdelapenya** [comment]
>Does that look fine for you?

Yes, we are not in a hurry at all, so please work on your own priorities.

> Are you using default values or overriding them?

We are just using the vanilla Helm chart, no overrides.

> Do you have other log to help debug?

About the logs, this is the only CI log that we have: https://apm-ci.elastic.co/blue/organizations/jenkins/stack%2Fe2e-testing-mbp/detail/PR-129/5/pipeline#step-262-log-474. I think we could try to improve the error logs though

**Comment by mdelapenya** [comment]
@jmlrt more on this: here it's an enhanced log error: https://apm-ci.elastic.co/blue/organizations/jenkins/stack%2Fe2e-testing-mbp/detail/PR-129/6/pipeline/202#step-296-log-469

> yaml: line 7: mapping values are not allowed in this context
```yaml
Name:              apm-server-apm-server
Namespace:         default
Labels:            app=apm-server
                   app.kubernetes.io/managed-by=Helm
                   heritage=Helm
                   release=apm-server
Annotations:       meta.helm.sh/release-name: apm-server
                   meta.helm.sh/release-namespace: default
Selector:          app=apm-server,release=apm-server
Type:              ClusterIP
IP:                10.106.243.93
Port:              http  8200/TCP
TargetPort:        8200/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
```

**Comment by jmlrt** [comment]
I don't see anything wrong in these values.

My local test seem to work and `kubectl describe service` result is similar to yours:

```
➜  helm-3-apm git:(master) kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-21T14:50:54Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:18:29Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}

➜  helm-3-apm git:(master) helm version          
version.BuildInfo{Version:"v3.2.3", GitCommit:"8f832046e258e2cb800894579b1b3b50c2d83492", GitTreeState:"clean", GoVersion:"go1.13.12"}

➜  helm-3-apm git:(master) helm install apm-server apm-server 
NAME: apm-server
LAST DEPLOYED: Wed Jun 10 11:34:41 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
1. Watch all containers come up.
  $ kubectl get pods --namespace=default -l app=apm-server-apm-server -w

➜  helm-3-apm git:(master) kubectl describe service apm-server-apm-server 
Name:              apm-server-apm-server
Namespace:         default
Labels:            app=apm-server
                   app.kubernetes.io/managed-by=Helm
                   heritage=Helm
                   release=apm-server
Annotations:       meta.helm.sh/release-name: apm-server
                   meta.helm.sh/release-namespace: default
Selector:          app=apm-server,release=apm-server
Type:              ClusterIP
IP:                10.107.132.36
Port:              http  8200/TCP
TargetPort:        8200/TCP
Endpoints:         
Session Affinity:  None
Events:            <none>
```

Were you able to reproduce it outside of CI?

**Comment by mdelapenya** [comment]
Hi @jmlrt very sorry for the delay, I finally could put hands back on this.

Yes, I can reproduce it locally, where a YAML file seems not correct:

>ERRO[0182] Could not unmarshal YAML                      error="yaml: line 7: mapping values are not allowed in this context" 

```yaml
Name:              apm-server-apm-server
Namespace:         default
Labels:            app=apm-server
                   app.kubernetes.io/managed-by=Helm
                   heritage=Helm
                   release=apm-server
Annotations:       meta.helm.sh/release-name: apm-server
                   meta.helm.sh/release-namespace: default
Selector:          app=apm-server,release=apm-server
Type:              ClusterIP
IP:                10.98.150.9
Port:              http  8200/TCP
TargetPort:        8200/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
```

which seems exactly the same output generated by `kubectl`.

I guess the problem is with the YAML parser. Look: https://play.golang.org/p/E5hA5QAaGdJ

**Comment by mdelapenya** [comment]
Mmm, after playing a little bit with YAML indents, and also seeing Github pigmenting the above YAML syntax incorrectly, I think the problem comes with the output, which is missing a `/n` to change context at the Annotations key.

Please compare the above playground (https://play.golang.org/p/E5hA5QAaGdJ) with this one (https://play.golang.org/p/6xuuuVbfKFE):
- the `Annotations` item has two children

I'm gonna try bumping my local version of `kubectl`, which was `1.17.3` to `1.18.6` and will post here results

File: e2e/_suites/helm/helm_charts_test.go
Comment (mdelapenya) [inline e2e/_suites/helm/helm_charts_test.go:131]: @kuisathaverat @jmlrt are those endpoints the same retrieved by `kubectl get endpoints -o json`? Maybe we can simplify the calculation using it instead of `kubectl describe`, which does not have a JSON/YAML output.

Comment (jmlrt) [reply e2e/_suites/helm/helm_charts_test.go:131]: @mdelapenya, sorry for the late answer. Indeed `kubectl describe endpoints <my_endpoint>` and `kubectl get endpoints <my_endpoint> -o json` provide the same details, so you can use `kubectl get` with json or yaml output.

Comment (mdelapenya) [reply e2e/_suites/helm/helm_charts_test.go]: Thanks for you feedback @jmlrt! I addressed it using JSON format and parsing it with a dynamic JSON processor (gabs)

**Comment by mdelapenya** [comment]
Cool, Helm tests are passing @kuisathaverat @jmlrt if you approve, then we can merge this, providing support for Helm 3

BTW, @jmlrt we have these Helm tests scheduled nightly, if you want a slack message to a channel on failures, please let us know. We are doing this for some teams in other test suites.

**Review by kuisathaverat — APPROVED** [review]

**Review by jmlrt — COMMENTED** [review]
LGTM, however Helm 3 support is still in [Beta](https://github.com/elastic/helm-charts/#helm-3-beta) and we may not be ready to pass it in GA and remove Helm 2 support in November.

**Comment by jmlrt** [comment]
> BTW, @jmlrt we have these Helm tests scheduled nightly, if you want a slack message to a channel on failures, please let us know. We are doing this for some teams in other test suites.

Where are these tests running? It could be interesting to have failure notifications sent to `#infra-release-notify` if the logs are relevant for us.

**Comment by mdelapenya** [comment]
> > BTW, @jmlrt we have these Helm tests scheduled nightly, if you want a slack message to a channel on failures, please let us know. We are doing this for some teams in other test suites.
> 
> Where are these tests running? It could be interesting to have failure notifications sent to `#infra-release-notify` if the logs are relevant for us.

We have a dedicated jenkins pipeline running helm tests nightly, under Beats CI. We could push a message to that channel on failures

File: cli/services/helm.go
Comment (mdelapenya) [reply cli/services/helm.go:147]: Agreed, we won't need it and a call to it is also not necessary, thanks for pointing out!

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:316]: @elastic/observablt-robots wdyt about this?

File: .ci/e2eTestingHelmDaily.groovy
Comment (mdelapenya) [inline .ci/e2eTestingHelmDaily.groovy:48]: @jmlrt after this is merged, you'll receive a notification when the Helm charts failed.

File: .ci/Jenkinsfile
Comment (mdelapenya) [inline .ci/Jenkinsfile:283]: After https://github.com/elastic/apm-pipeline-library/pull/803, the library will support multiple channels in the notify build result

**Comment by jmlrt** [comment]
@mdelapenya FYI, https://github.com/elastic/helm-charts has finally been updated to Helm 3 in https://github.com/elastic/helm-charts/pull/904. 7.10.0 for Helm charts will be released soon and include it.

**Comment by mdelapenya** [comment]
Great! We are running the tests nighty: https://beats-ci.elastic.co/blue/organizations/jenkins/e2e-tests%2Fe2e-testing-mbp/detail/master/722/pipeline/, in this case with with Helm 3.2.1 and Helm charts version 7.6.1

Feel free to update the versions in a PR if you like.

BTW, any failed nightly build will ping you on slack in the `#infra-release-notify` channel

**Comment by jmlrt** [comment]
> Feel free to update the versions in a PR if you like.

https://github.com/elastic/e2e-testing/pull/447

**Pull request merged**
